lib/module.ts

```diff
@@ -96,9 +68,43 @@ export class Module<
       flattenedModule[key] = subModules[key].flatten()
     }
 
+    // plugins
+    this.registerOnDataChangeHooks(state)
+
     return flattenedModule
   }
 
+  private getInitialState(): S | undefined {
+    const { stateInit, plugins } = this.options
+    const source = stateInit?.()
+    if (!plugins?.length) {
+      return source
+    }
+    return plugins.reduce((state, plugin) => {
+      if (plugin.onStateInit) {
+        return plugin.onStateInit(state ?? {}) as S
+      }
+      return state
+    }, source)
+  }
+
+  private registerOnDataChangeHooks(state: UnwrapNestedRefs<S>): void {
+    const onDataChangeHooks = this.options.plugins
+      ?.filter((x) => x.onDataChange !== undefined)
+      .map((x) => x.onDataChange)
+    if (onDataChangeHooks?.length) {
+      watch(
+        state,
+        (value, oldValue, onCleanup) => {
+          onDataChangeHooks.forEach((callback) => callback?.(value, oldValue, onCleanup))
+        },
+        {
+          immediate: true,
+        },
+      )
+    }
+  }
```

Comment:
I'm not sure if it's necessary to filter out those plugins without a `onDataChange` hook.
We can do it like the following too. which will make the code look cleaner:

```ts
watch(
  state,
  (value, oldValue, onCleanup) => {
    this.options.plugins?.forEach(plugin => plugin.onDataChange?.(value, oldValue, onCleanup))
  },
  {
    immediate: true,
  },
)
```

I think the amount of plugins may not be very high, so looping through all of them in `watch` might not be a problem.
Is there a better way to do this?

Reply:
I think the way you wrote it is cleaner. If the store doesn't need any plugins, it's nice to avoid the `watch` here.